framework/src/utils/PerfGraph.C

```diff
@@ -0,0 +1,107 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PerfGraph.h"
+
+// MOOSE Includes
+#include "PerfGuard.h"
+#include "MooseError.h"
+
+PerfGraph::PerfGraph() : _current_position(0), _active(true)
+{
+  // Not done in the initialization list on purpose because this object needs to be complete first
+  _root_node = libmesh_make_unique<PerfNode>(registerSection("root"));
+
+  _stack[0] = _root_node.get();
+
+  _root_start = std::chrono::steady_clock::now();
+}
+
+PerfGraph::~PerfGraph() {}
+
+unsigned int
+PerfGraph::registerSection(std::string section_name)
+{
+  auto it = _section_name_to_id.lower_bound(section_name);
+
+  // Is it already registered?
+  if (it != _section_name_to_id.lower_bound(section_name) && it->first == section_name)
+    mooseError("PerfGraph already holds a section named: ", section_name);
+
+  // It's not...
+  auto id = _section_name_to_id.size();
+  _section_name_to_id.emplace_hint(it, section_name, id);
+  _id_to_section_name[id] = section_name;
+
+  return id;
+}
+
+void
+PerfGraph::push(PerfID id)
+{
+  auto new_node = _stack[_current_position]->getChild(id);
+
+  _current_position++;
+
+  if (_current_position >= MAX_STACK_SIZE)
+    mooseError("PerfGraph is out of stack space!");
+
+  _stack[_current_position] = new_node;
+}
+
+void
+PerfGraph::pop(std::chrono::steady_clock::duration duration)
+{
+  _stack[_current_position]->addTime(duration);
```

Comment:
In libMesh, we had the ability to assert that the push and pops matched. If we only ever use scoped based timing, we won't need that, but as long as these are a public interface, I think that's a feature worth keeping even if only in devel/dbg modes (e.g. this would be a great thing to assert, but you'll need to change the `pop()` interface).